core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java

```diff
@@ -0,0 +1,218 @@
+
+    if (!emptyCipherStream) {
+      this.plainStreamPosition = 0;
+      this.fileAadPrefix = fileAadPrefix;
+      gcmDecryptor = new Ciphers.AesGcmDecryptor(aesKey);
+      plainBlockSize = ByteBuffer.wrap(prefixBytes, Ciphers.GCM_STREAM_MAGIC_ARRAY.length, 4)
+          .order(ByteOrder.LITTLE_ENDIAN).getInt();
+      Preconditions.checkState(plainBlockSize > 0, "Wrong plainBlockSize " + plainBlockSize);
+
+      cipherBlockSize = plainBlockSize + Ciphers.NONCE_LENGTH + Ciphers.GCM_TAG_LENGTH;
+      this.ciphertextBlockBuffer = new byte[cipherBlockSize];
+      this.currentBlockIndex = 0;
+      this.currentOffsetInPlainBlock = 0;
+
+      int numberOfFullBlocks = Math.toIntExact(netSourceFileSize / cipherBlockSize);
+      int cipherBytesInLastBlock = Math.toIntExact(netSourceFileSize - numberOfFullBlocks * cipherBlockSize);
+      boolean fullBlocksOnly = (0 == cipherBytesInLastBlock);
+      numberOfBlocks = fullBlocksOnly ? numberOfFullBlocks : numberOfFullBlocks + 1;
+      lastCipherBlockSize = fullBlocksOnly ? cipherBlockSize : cipherBytesInLastBlock; // never 0
+      int plainBytesInLastBlock = fullBlocksOnly ? 0 :
+          (cipherBytesInLastBlock - Ciphers.NONCE_LENGTH - Ciphers.GCM_TAG_LENGTH);
+      plainStreamSize = numberOfFullBlocks * plainBlockSize + plainBytesInLastBlock;
+    } else {
+      plainStreamSize = 0;
+
+      gcmDecryptor = null;
+      ciphertextBlockBuffer = null;
+      cipherBlockSize = -1;
+      plainBlockSize = -1;
+      numberOfBlocks = -1;
+      lastCipherBlockSize = -1;
+      this.fileAadPrefix = null;
+    }
+  }
+
+  public long plaintextStreamSize() {
+    return plainStreamSize;
+  }
+
+  @Override
+  public int available() throws IOException {
+    long maxAvailable = plainStreamSize - plainStreamPosition;
+    // See InputStream.available contract
+    if (maxAvailable >= Integer.MAX_VALUE) {
+      return Integer.MAX_VALUE;
+    } else {
+      return (int) maxAvailable;
+    }
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    if (len < 0) {
+      throw new IOException("Negative read length " + len);
+    }
+
+    if (available() <= 0) {
+      return -1;
+    }
+
+    boolean lastBlock = (currentBlockIndex + 1 == numberOfBlocks);
+    int resultBufferOffset = off;
+    int remaining = len;
+
+    sourceStream.seek(Ciphers.GCM_STREAM_PREFIX_LENGTH + currentBlockIndex * cipherBlockSize);
+
+    while (remaining > 0) {
+      int toLoad = lastBlock ? lastCipherBlockSize : cipherBlockSize;
+      int loaded = sourceStream.read(ciphertextBlockBuffer, 0, toLoad);
+      if (loaded != toLoad) {
+        throw new IOException("Should read " + toLoad + " bytes, but got only " + loaded + " bytes");
+      }
+
+      byte[] aad = Ciphers.streamBlockAAD(fileAadPrefix, currentBlockIndex);
+      byte[] plaintextBlock = gcmDecryptor.decrypt(ciphertextBlockBuffer, 0, toLoad, aad);
```

Comment:
It looks to me like this is going to read and decrypt the current block every time `read` is called. Am I wrong? Why not keep the current decrypted content in a buffer that persists across calls and consume bytes as needed from that buffer?

Reply:
Yep, the current code is focused on manifest files, so I thought of adding this optimization to a later version for heavier  Avro data files. But I can add a single-block buffer (for the "current position block") now, shouldn't be too much extra complexity. Later, we can consider multi-block buffers and their management.